### ISSUE TITLE
`FrechetInceptionDistance` hangs when used with multiple devices in PytorchLightning

### ISSUE THREAD

**Issue by will-rice** [Lightning-AI/torchmetrics#2684]
## 🐛 Bug

When used in the `on_validation_end` hook `FrechetInceptionDistance` hangs when using more than one device. I have `on_validation_end` decorated with `rank_zero_only` so I'm not sure why multiple devices would even affect this. However, other torchmetrics (SSIM, PSNR, and MS-SSIM) run right before and don't have any issues. One more thing is that I've run this code distributed before and it worked so there must have been an update that introduced this behavior.

### To Reproduce

Steps to reproduce the behavior...

<details>
  <summary>Code sample</summary>

```py
    @rank_zero_only
    def on_validation_end(self) -> None:
        """On validation end."""
        if self.global_rank == 0:
            video = self.last_val_batch["video"]
            identity = self.last_val_batch["identity"]
            audio = self.last_val_batch["audio"]
            masks = self.last_val_batch["masks"]
            overlapping = self.last_val_batch["overlapping"]

            with torch.no_grad():
                images_pred = self(
                    overlapping.flatten(0, 1),
                    video.flatten(0, 1),
                    masks.flatten(0, 1),
                    identity.flatten(0, 1),
                )
            images = utils.denormalize(video.flatten(0, 1))
            psnr = self.psnr(images_pred, images)
            ssim = self.ssim(images_pred, images)
            ms_ssim = self.ms_ssim(images_pred, images)
            self.fid.update(images_pred, real=False)
            self.fid.update(images, real=True)
            fid = self.fid.compute()

            wandb.log({"psnr": psnr, "ssim": ssim, "ms_ssim": ms_ssim, "fid": fid})
```

</details>

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source):
- Python & PyTorch Version (e.g., 1.0):
- Any other relevant information such as OS (e.g., Linux):

**Comment by Borda** [comment]
> When used in the `on_validation_end` hook `FrechetInceptionDistance` hangs when using more than one device

hello, just to clarify, multiple devices you mean multi-machine training (MMT)?

**Comment by SkafteNicki** [comment]
Hi @will-rice, thanks for raising this issue. I am pretty sure what goes wrong here is that torchmetrics will still try do a distributed sync when you call `self.fid.compute()`. Here is the reason:
* For all the other metrics you essentially call the `metric.forward` function. This function will by default not try to sync between all devices in the group because it would be very expensive to do so for every batch.
* For `FrechetInceptionDistance` you instead call `update` with `compute` (because `update` needs to be called with both positive and negative samples) and the default behavior of `compute` is to do sync between all devices in the group.

The sync in torchmetrics is currently by default done between all devices in the world:
https://github.com/Lightning-AI/torchmetrics/blob/a31417ca1bc4a47ca093c22e3f2f2b5e6843d58a/src/torchmetrics/utilities/distributed.py#L111-L112

This will basically ignore your `rank_zero_only` decorator and try to do a sync between all devices in the group, which leads to all processes to hang because of race conditions (only one process actually tries to sync and waits for the others)

The good news is that you can disable this behavior by initializing the metric as:
```python
from torchmetrics.image import FrechetInceptionDistance
fid = FrechetInceptionDistance(sync_on_compute=False)
```
which will disable the global sync that happens in `compute` method.

This may seem strange behavior, but it was a design decision taken in the early stages of torchmetrics to make the distributed synchronization as easy to use for the 99% of users. Sorry for the inconvenience this has caused. I have created PR #2719 that adds this particular pitfall to the documentation of what to be aware of when using torchmetrics with lightning:
https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html#common-pitfalls
